File: algos/stack/generate-parentheses.js
function generate(n) {
  let stack = [];
  let res = [];

  function backtrack(openNum, closeNum) {
    // console.log(stack)
    if (openNum === n && closeNum === n) {
      res.push(stack.join(""));
      return;
    }
    if (openNum < n) {
      stack.push("(");
      backtrack(openNum + 1, closeNum);
      stack.pop();
    }
    if (closeNum < openNum) {
      stack.push(")");
      backtrack(openNum, closeNum + 1);
      stack.pop();
    }
  }

  backtrack(0, 0);
  return res;
}

console.log(generate(3));

File: algos/stack/generate-parentheses.py
def generate(n):
    stack = []
    res = []

    def backtrack(open, close):
        print(stack)
        if open == close == n:
            res.append("".join(stack))
            return

        if open < n:
            stack.append("(")
            backtrack(open + 1, close)
            stack.pop()

        if close < open:
            stack.append(")")
            backtrack(open, close + 1)
            stack.pop()

    backtrack(0, 0)
    return res


print(generate(2))
